### PR TITLE
Fix ratslife crash

### LINKED ISSUE #3982: Crash of ratslife.wbt
**Describe the Bug**
`ratslife.wbt` crash when running the `contest_manager`

It seems to come from #3830.

### PR DESCRIPTION
**Pull request by BenjaminDeleze**
**Description**
The problem came from the fact that we were using the url to find the image in the `gImagesMap`. It is fine and handy to retrieve an image from a given url, but when we want to remove an image it does not work.

It does not work because the workflow goes as follow:

1. We/supervisor change the value of an url of a `imageTexture`.
2. The value of the url of the `imageTexture` is effectively changed.
3. A signal is emitted telling `imageTexture` that the url has changed.
4. `destroyWrenTexture` is called to clean the old image.
5. It was using the url to lookup the map. But the url was already the **new** url, so it was deleting the wrong image

Now it iterates through the images (an not through urls) when we want to delete an image.

Fix #3982

### CODE REVIEW

File: src/webots/nodes/WbImageTexture.cpp [hunk 1]
```diff
@@ -272,16 +272,21 @@ void WbImageTexture::destroyWrenTexture() {
...
+    if (image && image == mImage) {
+      QString key = i.key();
+      const int instances = i.value().second - 1;
+      assert(instances >= 0);
+      if (instances == 0) {
+        delete mImage;
+        gImagesMap.remove(key);
+      } else {
+        gImagesMap[key] = std::make_pair(image, instances);
+      }
```
Comment (omichel) [inline src/webots/nodes/WbImageTexture.cpp:288]: Coding style:
```suggestion
      } else
        gImagesMap[key] = std::make_pair(image, instances);
```

File: src/webots/nodes/WbImageTexture.cpp [hunk 2]
```diff
@@ -272,16 +272,21 @@ void WbImageTexture::destroyWrenTexture() {
...
-      delete mImage;
-      gImagesMap.remove(url);
-    } else
-      gImagesMap[url] = std::make_pair(pair.first, instances);
+  QMapIterator<QString, std::pair<QImage *, int>> i(gImagesMap);
+  while (i.hasNext()) {
+    i.next();
+    QImage *image = i.value().first;
+    if (image && image == mImage) {
+      QString key = i.key();
```
Comment (omichel) [inline src/webots/nodes/WbImageTexture.cpp:280]:
```suggestion
      const QString key = i.key();
```

File: src/webots/nodes/WbImageTexture.cpp [hunk 3]
```diff
@@ -272,16 +272,21 @@ void WbImageTexture::destroyWrenTexture() {
...
-    assert(instances >= 0);
-    if (instances == 0) {
-      delete mImage;
-      gImagesMap.remove(url);
-    } else
-      gImagesMap[url] = std::make_pair(pair.first, instances);
+  QMapIterator<QString, std::pair<QImage *, int>> i(gImagesMap);
+  while (i.hasNext()) {
+    i.next();
+    QImage *image = i.value().first;
```
Comment (omichel) [inline src/webots/nodes/WbImageTexture.cpp:278]:
```suggestion
    const QImage *image = i.value().first;
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Review by ad-daniel — APPROVED** [review]
Nice catch, no issues with the memory leak either. Thank you.

**Review by omichel — APPROVED** [review]
Looks good to me. Thank you.

**Pull request merged**